hoodie-common/src/main/java/com/uber/hoodie/common/HoodieFileGroupBuilder.java

```diff
@@ -0,0 +1,184 @@
+    // It is possible that the log-file was generated because of an inflight delta-instant. Even in this case, the log
+    // file will be allowed to added to the file-groups. The log-blocks(log-files) belonging to inflight delta-instants
+    // are filtered-out during record-reading.
+    if (!fileSlices.containsKey(logFile.getBaseCommitTime())) {
+      fileSlices.put(logFile.getBaseCommitTime(), new FileSlice(logFile.getBaseCommitTime(), id));
+    }
+    fileSlices.get(logFile.getBaseCommitTime()).addLogFile(logFile);
+    return this;
+  }
+
+  @Override
+  public HoodieFileGroup build() {
+    Preconditions.checkNotNull(id, "File Id must not be null");
+    Preconditions.checkNotNull(timeline, "Timeline must not be null");
+
+    // We can have at-most last 2 file-slices without base file. This is possible if a compaction request
+    // has been set for a file-group first time. In this case, the file group may not even have a data file
+    // (in the case of log files supporting inserts).
+    Preconditions.checkArgument(fileSlices.values().stream().filter(f -> !f.getDataFile().isPresent()).count() <= 2);
+    Map<String, FileSlice> mergedFileSlices = fileSlices;
+    if (!fileSlices.isEmpty()) {
+      FileSlice lastSlice = fileSlices.lastEntry().getValue();
+      // When file-group has more than one file-slice and the last file-slice does not have data-file, then
+      // it is the pending compaction case
+      if (!lastSlice.getDataFile().isPresent() && (fileSlices.size() > 1)) {
+        // last file slice is due to compaction (fake)
+        mergedFileSlices = adjustFileGroupSlices(fileSlices, lastSlice);
+      }
+    }
+    return new HoodieFileGroup(partitionPath, id, timeline, mergedFileSlices);
+  }
+
+  /**
+   * Helper to merge fake file-slice due to pending compaction with latest file-slice and provide the correct
+   * latest view of the file-group
+   *
+   * @param rawFileSlices File Slices including fake file slices
+   * @param fakeFileSlice Fake file slice
+   */
+  private static Map<String, FileSlice> adjustFileGroupSlices(NavigableMap<String, FileSlice> rawFileSlices,
+      FileSlice fakeFileSlice) {
+    final Map<String, FileSlice> mergedFileSlices = new TreeMap<>(HoodieFileGroup.getReverseCommitTimeComparator());
+    // case where last file-slice is due to compaction request.
+    // We need to merge the latest 2 file-slices
+    Entry<String, FileSlice> lastEntry = rawFileSlices.lastEntry();
+    Entry<String, FileSlice> penultimateEntry = rawFileSlices.lowerEntry(lastEntry.getKey());
+    Preconditions.checkArgument(lastEntry.getValue() == fakeFileSlice,
+        "Sanity check to ensure the last file-slice is the one not having the data-file. "
+            + "Last Entry=" + lastEntry.getValue()
+            + " Expected=" + fakeFileSlice);
+    FileSlice merged = mergeFileSlices(lastEntry.getValue(), penultimateEntry.getValue());
```

Comment:
Better naming for this please ? Like what is this merging ? mergedLastTwoFileSlicesWithoutBaseDataFile ?